### PR TITLE
KAK decomposition triggered only if basis gate count in a block is not equal to that of KAK 

### LINKED ISSUE #2415: Extra U3 gates inserted into circuits by ConsolidateBlocks pass
### Information

- **Qiskit Terra version**: master
- **Python version**:
- **Operating system**: 

### What is the current behavior?
These two decompositions should be the same:

```python
import numpy as np
from qiskit import *

q = QuantumRegister(2)
c = ClassicalRegister(2)
qc = QuantumCircuit(q, c)

qc.cz(q[0], q[1])
qc.barrier(q)
qc.cu1(np.pi, q[0], q[1])

cz_circ = transpile(qc, None, basis_gates=['u1', 'u2', 'u3', 'id', 'cx'])
cz_circ.draw()

```
Screen Shot 2019-05-14 at 3 49 57 AM

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary

Fixes #2415. 

The KAK block collection is triggered only if the unitary generated by a collected block has a different number of cx gates than the cx equivalent unitary, as determined by the local invariant parameters.

Note that KAK is triggered both if the block unitary cx count is above or below the cx equivalent.  This is because the former indicates that KAK can optimize the circuit, while the latter indicates that the circuit was not unrolled into the basis and we should KAK a try.

### CODE REVIEW

**Comment by levbishop** [comment]
The concept to avoid rewriting if the basis count won't change is good, but I think it's shortsighted to restrict only to exact CNOT decompositions since we now have the ability to do approximate decompositions and use other basis gates (with more to come in the future). I suggest that rather than duplicating such logic in the local_invariance.py just add a method to `TwoQubitBasisDecomposer` something like `decomposer.required_basis_uses(target_unitary)` that returns the number of basis uses needed for the given target, basis set, approximation error, etc, etc

**Comment by nonhermitian** [comment]
So the whole motivation was to get around the extra (possibly hundreds or more) U3 gates that get added to circuits by the KAK.  See for example: https://github.com/Qiskit/qiskit-terra/issues/2415#issuecomment-492671237

This method is hard coded for cx gates, or any gate locally equivalent to that, e.g. cz, but those are the entangling gates that we use.  It will let other stuff fly through if it sees a mismatch between the cx count and what is it locally equivalent to, so other bases of entangling gates will still be collected as before.

The transpiler has already decided to lump the gates into a single unitary by the time the KAK decomposition routines are called, so it is too late at that point to stop the extra U3 gates from being added to the circuit.

I do not think that this is duplicate logic.  This is much cheaper to compute than Weyl chamber coordinates, especially the way they are computed now, so it is an efficient way to determine whether or not there is any worth in decomposition.  This is why it is where it is in the block consolidation pass.

**Comment by levbishop** [comment]
I mean instead of the test being 
```python
cx_equivalence(unitary.to_matrix()) != cx_count
``` 
it could be 
```python
decomposer.basis_equivalence(unitary) != basis_count
```
Where decomposer is an instance of `TwoQubitDecomposer` and knows for example if we are doing an approximate decomposition and can drop more CNOTs because of that.

The implementation of `TwoQubitDecomposer.basis_equivalence()` could still use this Makhlin calculation of the Weyl coordinates for speed if needed.

**Comment by nonhermitian** [comment]
This now uses the Weyl coordinates themselves to get the basis gate equivalence via the `TwoQubitBasisDecomposer`.  This required adding a basis gate kwarg to the `ConsolidateBlocks` pass, but it now allows for any gate to be used as the decomposition check.

File: test/python/transpiler/test_kak_over_optimization.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                            basis_gates=['u1', 'u2', 'u3', 'id', 'cx'],
+                            optimization_level=3)
+        ops = cz_circ.count_ops()
+        self.assertEqual(ops['u2'], 2)
+        self.assertEqual(ops['cx'], 1)
+        self.assertFalse('u3' in ops.keys())
+
+    def test_cu1_optimization(self):
+        """ Test that KAK does run on a cu1 gate and
+        reduces the cx count from two to one.
```
Comment (mtreinish) [inline test/python/transpiler/test_kak_over_optimization.py]: I ran this test locally without your patch applied and was not able to reproduce a case with 2 cx gates in the output ops. It always only had 1 cx listed.

**Comment by nonhermitian** [comment]
Yes, I  think it is because our decomposition rules have been improved (#2600 I believe). It should not be two cx, but the old rules were not smart enough to know that.  However, the more important of the two tests is `test_cz_optimization` since that used to generate a bunch of `u3` gates, which is the error fixed here.

**Pull request merged**